homeassistant/components/ecoforest/sensor.py

```diff
@@ -0,0 +1,83 @@
+"""Support for IPMA sensors."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+import logging
+
+from pyecoforest.models.device import Device
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import UnitOfTemperature
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+from .coordinator import EcoforestCoordinator
+from .entity import EcoforestEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class EcoforestRequiredKeysMixin:
+    """Mixin for required keys."""
+
+    value_fn: Callable[[Device], float | None]
+
+
+@dataclass
+class EcoforestSensorEntityDescription(
+    SensorEntityDescription, EcoforestRequiredKeysMixin
+):
+    """Describes Ecoforest sensor entity."""
+
+
+SENSOR_TYPES: tuple[EcoforestSensorEntityDescription, ...] = (
+    EcoforestSensorEntityDescription(
+        key="temperature",
+        translation_key="temperature",
+        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+        device_class=SensorDeviceClass.TEMPERATURE,
+        value_fn=lambda data: data.environment_temperature,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Ecoforest sensor platform."""
+    coordinator: EcoforestCoordinator = hass.data[DOMAIN][entry.entry_id]
+    assert coordinator.data is not None
+
+    entities = [
+        EcoforestSensor(coordinator, description) for description in SENSOR_TYPES
+    ]
+
+    async_add_entities(entities, False)
+
+
+class EcoforestSensor(SensorEntity, EcoforestEntity):
+    """Representation of an Ecoforest sensor."""
+
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        coordinator: EcoforestCoordinator,
+        description: EcoforestSensorEntityDescription,
+    ) -> None:
+        """Initialize the Ecoforest Sensor."""
+        self.description = description
```

Comment:
You're already doing this in the entity. You can make sure the types work to declare `entity_description: EcoforestSensorEntityDescription` outside of the constructor

Reply:
I am learning python while doing this, so my apologies my language knowledge is limited. I tried to work around this and wasn't able to.

```python
class EcoforestSensor(SensorEntity, EcoforestEntity):
    """Representation of an Ecoforest sensor."""

    def __init__(
        self,
        coordinator: EcoforestCoordinator,
        description: EcoforestSensorEntityDescription,
    ) -> None:
        """Initialize the Ecoforest Sensor."""
        entity_description: EcoforestSensorEntityDescription

        EcoforestEntity.__init__(self, coordinator, description)

    @property
    def native_value(self) -> float | None:
        """Return the state of the sensor."""
        return self.entity_description.value_fn(self.data)
```
While using this I get a ruff error with:
```shell 
homeassistant/components/ecoforest/sensor.py:73:9: F842 Local variable `entity_description` is annotated but never used
Found 1 error.
```

and  mypy error with:
```shell 
homeassistant/components/ecoforest/sensor.py:80: error: "SensorEntityDescription" has no attribute "value_fn"  [attr-defined]
Found 1 error in 1 file (checked 1 source file)
```

Reply:
Outside of the constructor just like in 
https://github.com/home-assistant/core/blob/6752af8f27516be52600c1782f160d724760e869/homeassistant/components/youtube/sensor.py#L87